test/testlib/multicluster_utilities.go

```diff
@@ -243,3 +253,97 @@ func AdjustPodTimeout(podName string, timeout time.Duration) time.Duration {
 	}
 	return timeout
 }
+
+const (
+	DNS_CONFIG_MARKER = "# BEGIN BASE CONFIG"
+	COREDNS_NS        = "kube-system"
+	COREDNS_CM        = "coredns"
+	COREFILE_KEY      = "Corefile"
+	DNS_SERVICE       = "dns-service"
+)
+
+func getClusterDnsServer(t *testing.T, clusterContext context.Context) string {
+	var ret string
+	DeployWithContext(t,
+		clusterContext,
+		func(context *ClusterDeploymentContext, options *helm.Options) {
+			// Copy kubectl options and update namespace
+			require.NotNil(t, options.KubectlOptions)
+			kubectlOptions := *options.KubectlOptions
+			kubectlOptions.Namespace = COREDNS_NS
+			// Get endpoint for DNS server
+			out, err := k8s.RunKubectlAndGetOutputE(t, &kubectlOptions,
+				"get", "service", DNS_SERVICE,
+				"-o", "jsonpath={.status.loadBalancer.ingress[0].hostname}")
+			require.NoError(t, err)
+			ret = strings.TrimSpace(out)
+		},
+	)
+	return ret
+}
+
+func getDnsServerIps(t *testing.T, clusterContext context.Context) []string {
+	host := getClusterDnsServer(t, clusterContext)
+	if host == "" {
+		return nil
+	}
+	addresses, err := net.LookupHost(host)
+	require.NoError(t, err)
+	sort.Strings(addresses)
+	return addresses
+}
+
+func getDnsConfigSnippet(t *testing.T, context context.Context) string {
+	dnsServers := getDnsServerIps(t, context)
+	if dnsServers == nil {
+		return ""
+	}
+	deploymentContext := GetClusterDeploymentContext(t, context)
+	return fmt.Sprintf(`%s:53 {
+    errors
+    cache 30
+    forward . %s {
+      force_tcp
+    }
+}`, deploymentContext.ThisCluster.Domain, strings.Join(dnsServers, " "))
+}
+
+func updateDnsConfig(t *testing.T, ctx context.Context, kubectlOptions *k8s.KubectlOptions, dnsConfigSnippet string) {
+	// Create K8s client and get configmap for CoreDNS
+	clientset, err := k8s.GetKubernetesClientFromOptionsE(t, kubectlOptions)
+	require.NoError(t, err, "Unable to create K8s client")
+	cm, err := clientset.CoreV1().ConfigMaps(COREDNS_NS).Get(ctx, COREDNS_CM, v1.GetOptions{})
+	require.NoError(t, err, "Unable to get CoreDNS configmap")
+	config, ok := cm.Data[COREFILE_KEY]
+	require.True(t, ok, "Did not find key %s in CoreDNS configmap", COREFILE_KEY)
+
+	// Find marker string separating generated config from base config
+	idx := strings.Index(config, DNS_CONFIG_MARKER)
+	require.GreaterOrEqual(t, idx, 0, "Did not find marker string in CoreDNS config")
+
+	// Update configmap if generated config changed
+	updatedConfig := dnsConfigSnippet + "\n" + config[idx:]
+	if updatedConfig == config {
+		t.Logf("DNS configuration up to date")
+		return
+	}
+	t.Logf("Adding DNS config snippet %s", dnsConfigSnippet)
+	cm.Data[COREFILE_KEY] = updatedConfig
+	_, err = clientset.CoreV1().ConfigMaps(COREDNS_NS).Update(ctx, cm, v1.UpdateOptions{})
```

Comment:
Does this require a CoreDNS service restart? I remember that I needed to restart the pods after changing the `ttl` value in the same configuration.

Reply:
I see that the CoreDNS containers in the EKS have not undergone any restarts and they seem to dynamically pick up changes to the config.

```console
$ kubectl -n kube-system get pods
NAME                                 READY   STATUS    RESTARTS         AGE
...
coredns-7454b7b4d5-s5w9v             1/1     Running   0                21d
coredns-7454b7b4d5-xcfxn             1/1     Running   0                21d
...
```

It looks like CoreDNS reloads the config when it detects changes to the hash of the config.

```console
$ kubectl -n kube-system logs coredns-7454b7b4d5-xcfxn
...
[INFO] plugin/reload: Running configuration SHA512 = 5433e9ae056a43e2582ccea14671ad5dccbab0a40f9f731144af4f8a6b33066dd10f78d2d8926ebfadebd4c2b0800a583581386c3f97c21aa032d58c3e029507
[INFO] Reloading complete
```

Reply:
Thanks for checking!